File: medium/typescript/155-min-stack.ts
class MinStack {
  private stack: number[] = [];
  private min: number[] = [];

  push(val: number): void {
    const min = this.min.length === 0 ? val : Math.min(val, this.getMin());

    this.stack.push(val);
    this.min.push(min);
  }

  pop(): void {
    this.stack.pop();
    this.min.pop();
  }

  top(): number {
    return this.stack[this.stack.length - 1];
  }

  getMin(): number {
    return this.min[this.min.length - 1];
  }
}

const stack = new MinStack();

File: medium/python/155-min-stack.py
class MinStack:
    def __init__(self):
        self.min: list[int] = []
        self.stack: list[int] = []

    def push(self, val: int) -> None:
        self.stack.append(val)
        val = min(val, self.getMin() if self.min else val)
        self.min.append(val)

    def pop(self) -> None:
        self.stack.pop()
        self.min.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.min[-1]
